middleware/rewrite/edns0.go

```diff
@@ -0,0 +1,132 @@
+// Package rewrite is middleware for rewriting requests internally to something different.
+package rewrite
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/miekg/dns"
+)
+
+// edns0LocalRule is a class rewrite rule.
+type edns0LocalRule struct {
+	action string
+	code   uint16
+	data   []byte
+	match  []byte // TODO: add match input for replace rule
+}
+
+type edns0NsidRule struct {
+	action string
+}
+
+func getEdns0Opt(r *dns.Msg) *dns.OPT {
+	o := r.IsEdns0()
+	if o == nil {
+		r.SetEdns0(4096, true)
+		o = r.IsEdns0()
+	}
+	return o
+}
+
+func (rule *edns0NsidRule) Rewrite(r *dns.Msg) Result {
+	o := getEdns0Opt(r)
+	found := false
+	for _, s := range o.Option {
+		switch e := s.(type) {
+		case *dns.EDNS0_NSID:
+			if rule.action == "replace" || rule.action == "set" {
+				e.Nsid = "" // make sure it is empty for request
+			}
+			found = true
+			break
+		}
+	}
+
+	// add option if not found
+	if !found && (rule.action == "append" || rule.action == "set") {
+		o.SetDo(true)
+		o.Option = append(o.Option, &dns.EDNS0_NSID{dns.EDNS0NSID, ""})
+	}
+
+	return RewriteDone
+}
+
+// SetEDNS0Attrs will alter the request
+func (rule *edns0LocalRule) Rewrite(r *dns.Msg) Result {
+	o := getEdns0Opt(r)
+	found := false
+	for _, s := range o.Option {
+		switch e := s.(type) {
+		case *dns.EDNS0_LOCAL:
+			if rule.code == e.Code {
+				if rule.action == "replace" || rule.action == "set" {
+					e.Data = rule.data
+				}
+				found = true
+				break
+			}
+		}
+	}
+
+	// add option if not found
+	if !found && (rule.action == "append" || rule.action == "set") {
+		o.SetDo(true)
+		var opt dns.EDNS0_LOCAL
+		opt.Code = rule.code
+		opt.Data = rule.data
+		o.Option = append(o.Option, &opt)
+	}
+
+	return RewriteDone
+}
+
+// Initializer
+func newEdns0Rule(args ...string) (Rule, error) {
+	if len(args) < 2 {
```

Comment:
This one still needs varargs, because it will create different EDNS0 rules depending on the args, and those need different number of args. However, they all require at least 2 args: the action, and the EDNS0 option type.